### ISSUE TITLE
ROCm 5.4.2 + Pytorch seems not working correctly on ubuntu 20.04.1

### ISSUE THREAD

**Issue by mapreduce1314** [RadeonOpenCompute/ROCm#2100]
Environments:
Hardware: RX6600 XT
1 Ubuntu - 5.15.0-72-generic #79~20.04.1-Ubuntu SMP
2 Installed dkms, rocms by amdgpu-install 5.4.2, following this guidance:
https://docs.amd.com/bundle/ROCm-Installation-Guide-v5.4.2/page/How_to_Install_ROCm.html
3 Python 3.10.10
4 Install pytorch stable 2.0.0 for ROCm 5.4.2 from https://pytorch.org/
5 Already export HSA_OVERRIDE_GFX_VERSION=10.3.0 

The GPU seems to work but if I look at the loss function result of a simple Linear Model, I find that the loss is quite different between CPU & GPU result. CPU loss is 20.36 while GPU is -2.26.

The negative loss seems not correct when running the simple model in GPU, but no sure which is the direction to try, switch to another pytorch  + ROCm version set or some simple configuration change? Please help!

Detailed information:

The program output is:

------
CPU model weights: Parameter containing:
tensor([[0.9534, 1.0810]], requires_grad=True)
GPU model weights: Parameter containing:
tensor([[0.9534, 1.0810]], requires_grad=True)
CPU loss: 20.36042022705078
GPU loss: -1.7142915703512032e+38

Weight difference: tensor([[0.0000e+00, 1.1921e-07]], grad_fn=<AbsBackward0>)
Loss difference: 1.7142915703512032e+38

-----

Full python code as follows:

import torch
import torch.nn as nn
import torch.optim as optim

torch.manual_seed(42)

class LinearModel(nn.Module):
    def __init__(self):
        super(LinearModel, self).__init__()
        self.fc = nn.Linear(2, 1)

    def forward(self, x):
        return self.fc(x)

input_data = torch.tensor([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0], [7.0, 8.0]], dtype=torch.float32)
target_data = torch.tensor([[3.0], [7.0], [11.0], [15.0]], dtype=torch.float32)

cpu_model = LinearModel()
gpu_model = LinearModel().to("cuda:0")

gpu_model.load_state_dict(cpu_model.state_dict())

cpu_optimizer = optim.SGD(cpu_model.parameters(), lr=0.01)
gpu_optimizer = optim.SGD(gpu_model.parameters(), lr=0.01)

criterion = nn.MSELoss()

cpu_optimizer.zero_grad()
cpu_output = cpu_model(input_data)
cpu_loss = criterion(cpu_output, target_data)
cpu_loss.backward()
cpu_optimizer.step()

input_data_gpu = input_data.to("cuda:0")
target_data_gpu = target_data.to("cuda:0")
gpu_optimizer.zero_grad()
gpu_output = gpu_model(input_data_gpu)
gpu_loss = criterion(gpu_output, target_data_gpu)
gpu_loss.backward()
gpu_optimizer.step()

gpu_model.to("cpu")

print("CPU model weights:", cpu_model.fc.weight)
print("GPU model weights:", gpu_model.fc.weight)
print("CPU loss:", cpu_loss.item())
print("GPU loss:", gpu_loss.item())

weight_difference = torch.abs(cpu_model.fc.weight - gpu_model.fc.weight)
loss_difference = abs(cpu_loss.item() - gpu_loss.item())

print("\nWeight difference:", weight_difference)
print("Loss difference:", loss_difference)

rocminfo

HSA System Attributes    

Runtime Version:         1.1
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             

          
HSA Agents               
           
*******                  
Agent 1                  
*******                  
  Name:                    Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz
  Uuid:                    CPU-XX                             
  Marketing Name:          Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0(0x0)                             
  Queue Min Size:          0(0x0)                             
  Queue Max Size:          0(0x0)                             
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768(0x8000) KB                   
  Chip ID:                 0(0x0)                             
  ASIC Revision:           0(0x0)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   5100                               
  BDFID:                   0                                  
  Internal Node ID:        0                                  
  Compute Unit:            16                                 
  SIMDs per CU:            0                                  
  Shader Engines:          0                                  
  Shader Arrs. per Eng.:   0                                  
  WatchPts on Addr. Ranges:1                                  
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: FINE GRAINED        
      Size:                    16319152(0xf902b0) KB              
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    16319152(0xf902b0) KB              
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 3                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    16319152(0xf902b0) KB              
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
  ISA Info:                
*******                  
Agent 2                  
*******                  
  Name:                    gfx1030                            
  Uuid:                    GPU-XX                             
  Marketing Name:          AMD Radeon RX 6600 XT              
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128(0x80)                          
  Queue Min Size:          64(0x40)                           
  Queue Max Size:          131072(0x20000)                    
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      16(0x10) KB                        
    L2:                      2048(0x800) KB                     
    L3:                      32768(0x8000) KB                   
  Chip ID:                 29695(0x73ff)                      
  ASIC Revision:           0(0x0)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   2900                               
  BDFID:                   1536                               
  Internal Node ID:        1                                  
  Compute Unit:            32                                 
  SIMDs per CU:            2                                  
  Shader Engines:          4                                  
  Shader Arrs. per Eng.:   2                                  
  WatchPts on Addr. Ranges:4                                  
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      TRUE                               
  Wavefront Size:          32(0x20)                           
  Workgroup Max Size:      1024(0x400)                        
  Workgroup Max Size per Dimension:
    x                        1024(0x400)                        
    y                        1024(0x400)                        
    z                        1024(0x400)                        
  Max Waves Per CU:        32(0x20)                           
  Max Work-item Per CU:    1024(0x400)                        
  Grid Max Size:           4294967295(0xffffffff)             
  Grid Max Size per Dimension:
    x                        4294967295(0xffffffff)             
    y                        4294967295(0xffffffff)             
    z                        4294967295(0xffffffff)             
  Max fbarriers/Workgrp:   32                                 
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    8372224(0x7fc000) KB               
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 2                   
      Segment:                 GROUP                              
      Size:                    64(0x40) KB                        
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Alignment:         0KB                                
      Accessible by all:       FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx1030         
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        4294967295(0xffffffff)             
        y                        4294967295(0xffffffff)             
        z                        4294967295(0xffffffff)             
      FBarrier Max Size:       32                                 
*** Done ***             

rocm-smi
GPU  Temp (DieEdge)  AvgPwr  SCLK    MCLK   Fan  Perf  PwrCap  VRAM%  GPU%  
0    35.0c           4.0W    800Mhz  96Mhz  0%   auto  135.0W    5%   1%

**Comment by Dunedin87** [comment]
Hello, your issue is related to mine as well, where I was getting really odd things happening like negative losses (checked proper data and targets) and outputs being more than 1 or less than 0 even if it was softmax. The issue I think is Pytorch 2.0, I don't get the same issue in Pytorch 1.3. I was thinking of opening an issue in Pytorch's github page and I think you should too.

https://github.com/RadeonOpenCompute/ROCm/issues/2055

**Comment by mapreduce1314** [comment]
> Hello, your issue is related to mine as well, where I was getting really odd things happening like negative losses (checked proper data and targets) and outputs being more than 1 or less than 0 even if it was softmax. The issue I think is Pytorch 2.0, I don't get the same issue in Pytorch 1.3. I was thinking of opening an issue in Pytorch's github page and I think you should too.
> 
> #2055

Thanks for the input, would try to switch to Pytorch 1.3
I tried switching to tensorflow with ROCm, a similar model runs well on tensorflow. 
Looks like it is issue of Pytorch not ROCm, would close this one
